subsys/net/ip/tcp.c

```diff
@@ -748,8 +758,10 @@ static int tcp_update_recv_wnd(struct tcp *conn, int32_t delta)
 	bool short_win_after;
 
 	new_win = conn->recv_win + delta;
-	if (new_win < 0 || new_win > UINT16_MAX) {
```

Comment:
Did you deliberately allowed the new_win below 0, did you experience an occurance of this? Did you thought of how to handle the fact that new_win can have a value bigger then UINT16_MAX, but the implementation does no window scaling, so has no means to notify the other end of the connection about this?

Reply:
> Did you deliberately allowed the new_win below 0, did you experience an occurance of this?

This is to cover the scenario when the application shrinks the max window when some communication is already ongoing. In this case, the new current new window size could fall below 0.
At the same time, if the application updates the window after consuming the data, immediately after shrinking the new window size could exceed the max window, hence extra check.

> Did you thought of how to handle the fact that new_win can have a value bigger then UINT16_MAX, but the implementation does no window scaling, so has no means to notify the other end of the connection about this?

Not sure I follow. `new_win` could not exceed UINT16_MAX, as `recv_win_max` is `uint16_t`, and we make sure not to exceed it just below.

Reply:
> > Did you thought of how to handle the fact that new_win can have a value bigger then UINT16_MAX, but the implementation does no window scaling, so has no means to notify the other end of the connection about this?
> 
> Not sure I follow. `new_win` could not exceed UINT16_MAX, as `recv_win_max` is `uint16_t`, and we make sure not to exceed it just below.

The function tcp_update_recv_wnd is fed with an int32_t as well as the value new_win is. In current situation this could lead to an unexpected overflow when writing new_win to conn->recv_win. It might be worth looking at the origin code of this section

Reply:
I may be missing something, but aren't the current checks enough to guarantee that `new_win` ends up within the `uint16_t` range?

Reply:
You are right as recv_win_max is uint16_t this will be the limiting factor indeed, I did not spot this.